src/main/java/hudson/remoting/Engine.java

```diff
@@ -750,68 +774,73 @@ public void closeRead() throws IOException {
+                                        .put(
+                                                ClientProperties.PROXY_HEADERS,
+                                                Map.of(
+                                                        "Proxy-Authorization",
+                                                        "Basic "
+                                                                + Base64.getEncoder()
+                                                                        .encodeToString(proxyCredentials.getBytes(
+                                                                                StandardCharsets.UTF_8))));
+                            }
                         }
-                    }
 
-                    if (sslContext != null) {
-                        SslEngineConfigurator sslEngineConfigurator = new SslEngineConfigurator(sslContext);
-                        if (hostnameVerifier != null) {
-                            sslEngineConfigurator.setHostnameVerifier(hostnameVerifier);
+                        if (sslContext != null) {
+                            SslEngineConfigurator sslEngineConfigurator = new SslEngineConfigurator(sslContext);
+                            if (hostnameVerifier != null) {
+                                sslEngineConfigurator.setHostnameVerifier(hostnameVerifier);
+                            }
+                            client.getProperties().put(ClientProperties.SSL_ENGINE_CONFIGURATOR, sslEngineConfigurator);
                         }
-                        client.getProperties().put(ClientProperties.SSL_ENGINE_CONFIGURATOR, sslEngineConfigurator);
+                    }
+                    if (!succeedsWithRetries(() -> this.pingSuccessful(sslContext))) {
+                        return;
+                    }
+                    if (!succeedsWithRetries(() -> {
+                        container.connectToServer(
+                                new AgentEndpoint(),
+                                ClientEndpointConfig.Builder.create()
+                                        .configurator(headerHandler)
+                                        .build(),
+                                URI.create(wsUrl + "wsagents/"));
+                        return true;
+                    })) {
+                        return;
+                    }
+                    while (ch.get() == null) {
+                        Thread.sleep(100);
+                    }
+                    this.protocolName = "WebSocket";
+                    events.status("Connected");
+                    ch.get().join();
+                    events.status("Terminated");
+                    if (noReconnect) {
+                        return;
+                    }
+                    events.onDisconnect();
+                    reconnect();
+                } finally {
+                    if (container instanceof BaseContainer baseContainer) {
+                        baseContainer.shutdown();
```

Comment:
Unfortunately this only shuts down the main thread pool. Tyrus seems to maintain another thread pool in `org.glassfish.tyrus.container.jdk.client.TransportFilter.channelGroup` which I cannot figure out how to close; even tracking down which fields hold on to this object is challenging in such a complex library. This pool seems to have been overlooked by https://github.com/eclipse-ee4j/tyrus/pull/927.